### DISCUSSION TITLE
Obtain eval type of an expression [question]

### DISCUSSION THREAD

**Discussion by wolfv** [mitsuhiko/minijinja#302] [Ideas]
We are interested in being able to get the evaluation type of an expression, and were wondering if that is somehow possible.

E.g. with an expression like `{{ 5 if win else 3 }}` the eval type would be integer, etc. 

This could be useful for type-checking when we don't know which branch is being taken.

It would be nice if, in addition to `.undeclared_variables` the expression could have an `eval_type(context)` function that computes the type of the evaluation for all branches.

I also completely understand if that's out of scope for Minijinja :)

**Comment by mitsuhiko** [comment]
The issue is that the expression won't be known until it runs. While in some simple cases you can probably derive the type of the result, but in many cases you can't. In particular the moment you apply a filter or anything like this you cannot know the result.

It would be interesting what you are trying to accomplish. Maybe it can be accomplished in different ways.

**Comment by wolfv** [comment]
We have some "recipes" defined in YAML that we parse with a mix of serde_yaml + Minijinja.

There are sections like:

```
build:
  number: ${{ 0 if win else 100 }}
  string: ${{ "bla" if linux else "whatever" }}
```

From the serde schema we know that e.g. `build.number` is supposed to be an integer. It would be great if we could somehow infer that both arms of the `if` expression evaluate to integers as well (since locally we might only run into one of the two branches). If one of the two arms evaluates to a string we could warn or error out. 

I think for simple cases like this I can probably achieve it with some AST walking, and I agree, thinking about it a bit more it's really tough for everything that minijinja can do.

**Reply by mitsuhiko** [reply]
Yeah I think that might be impossible to accommodate.
